semantic-analysis/src/Analysis/Typecheck.hs

```diff
@@ -21,154 +22,144 @@ module Analysis.Typecheck
+import           Data.Semigroup (Last(..))
 import qualified Data.Set as Set
-import           Data.Traversable (for)
 import           Data.Void
 import           GHC.Generics (Generic1)
 import           Prelude hiding (fail)
-import           Source.Span
-import qualified Syntax.Algebra as Syntax
-import           Syntax.Functor
-import           Syntax.Module
-import           Syntax.Scope
-import qualified Syntax.Sum as Syntax
-import           Syntax.Term
-import           Syntax.Var (closed)
-import qualified System.Path as Path
-
-data Monotype f a
-  = Bool
+
+data Monotype a
+  = Var a
   | Unit
+  | Bool
+  | Int
   | String
-  | f a :-> f a
-  | Record (Map.Map Name (f a))
-  deriving (Foldable, Functor, Generic1, Traversable)
+  | Monotype a :-> Monotype a
+  -- | (Locally) undefined names whose types are unknown. May not be eliminated by unification.
+  | TypeOf Name
+  deriving (Eq, Foldable, Functor, Generic1, Ord, Show, Traversable)
 
 infixr 0 :->
 
-type Type = Term Monotype Meta
-
-type Addr = Name
-
 -- FIXME: Union the effects/annotations on the operands.
 
 -- | We derive the 'Semigroup' instance for types to take the second argument. This is equivalent to stating that the type of an imperative sequence of statements is the type of its final statement.
-deriving via (Last (Term Monotype a)) instance Semigroup (Term Monotype a)
+deriving via (Last (Monotype a)) instance Semigroup (Monotype a)
+
+instance Applicative Monotype where
+  pure = Var
+  (<*>) = ap
+
+instance Monad Monotype where
+  m >>= f = case m of
+    Var a    -> f a
+    Unit     -> Unit
+    Bool     -> Bool
+    Int      -> Int
+    String   -> String
+    TypeOf n -> TypeOf n
+    a :-> b  -> (a >>= f) :-> (b >>= f)
+
 
-deriving instance (Eq   a, forall a . Eq   a => Eq   (f a), Monad f) => Eq   (Monotype f a)
-deriving instance (Ord  a, forall a . Eq   a => Eq   (f a)
-                         , forall a . Ord  a => Ord  (f a), Monad f) => Ord  (Monotype f a)
-deriving instance (Show a, forall a . Show a => Show (f a))          => Show (Monotype f a)
+type Type = Monotype Meta
 
-instance HFunctor Monotype
-instance RightModule Monotype where
-  Unit      >>=* _ = Unit
-  Bool      >>=* _ = Bool
-  String    >>=* _ = String
-  (a :-> b) >>=* f = a >>= f :-> b >>= f
-  Record m  >>=* f = Record ((>>= f) <$> m)
 
 type Meta = Int
 
-newtype Polytype f a = PForAll (Scope () f a)
-  deriving (Foldable, Functor, Generic1, HFunctor, RightModule, Traversable)
 
-deriving instance (Eq   a, forall a . Eq   a => Eq   (f a), Monad f) => Eq   (Polytype f a)
-deriving instance (Ord  a, forall a . Eq   a => Eq   (f a)
-                         , forall a . Ord  a => Ord  (f a), Monad f) => Ord  (Polytype f a)
-deriving instance (Show a, forall a . Show a => Show (f a))          => Show (Polytype f a)
+data Polytype a
+  = PForAll (Polytype (Maybe a))
+  | PType (Monotype a)
+  deriving (Eq, Foldable, Functor, Generic1, Ord, Show, Traversable)
```

Comment:
What’d I tell you? A right `Type`-module.

cf: https://twitter.com/rob_rix/status/1146602548616413184